workspace/src/localization/localization_py/localization_py/EKF.py

```diff
@@ -0,0 +1,174 @@
+                ret = 1/(20*v)#-(self.tau_0*v/(self.omega_0*self.r_wheel*self.gamma))+self.tau_0
+        #print(ret)
+        if(ret<0):
+            return 0
+        return ret
+
+    def f_1_v(self, v):
+        return self.tau_0-v*self.tau_0/(self.omega_0*self.r_wheel*self.gamma)
+
+    def f_0_v(self, v):
+        if(v <= 0.1*self.omega_0*self.r_wheel):
+            return 0
+        else:
+            return self.tau_0/9-10*self.tau_0*v/(9*self.omega_0*self.gamma*self.r_wheel)
+    #Need to write this F
+
+    def new_calc_F(self, v, theta, delta, throttle):
+
+        F = np.array([
+            [self.dt*1.0, 0.0, - self.dt*v *
+                math.sin(theta), self.dt*math.cos(theta)],
+            [0.0, 1.0, self.dt*v *
+                math.cos(theta), self.dt*math.sin(theta)],
+            [0.0, 0.0, self.dt*1.0, self.dt*np.tan(delta)/self.l],
+            [0.0, 0.0, 0.0,
+            self.dt*(1+(self.r_wheel*self.gamma/(self.i_wheel))*(-(self.tau_0/(self.omega_0*self.r_wheel*self.gamma))-(self.c_1/(self.r_wheel*self.gamma))))]
+         ])
+
+        return F
+
+    def df_0_dv(self, v):
+        if(v <= 0.1*self.omega_0*self.r_wheel):
+            return 0
+        else:
+            return -10*self.tau_0 / (9*self.omega_0*self.r_wheel*self.gamma)
+    def __init__(self, dt):#, q1, q3, q4, r1, r3):
+        self.dt = dt
+        #vehicle parameters:
+        self.c_1 = 1e-4
+        self.c_0 = 0.02#0.039
+        self.l = 0.5
+        self.r_wheel = 0.08451952624
+        self.i_wheel = 1e-3
+        self.gamma = 1/3
+        self.tau_0 = 0.3#0.09
+        self.omega_0 = 30#161.185
+        self.df_1_dv = -self.tau_0/(self.omega_0*self.r_wheel*self.gamma)
+
+
+        # self.Q = np.diag([
+        #     q1,  # variance of location on x-axis
+        #     11,  # variance of location on y-axis
+        #     np.deg2rad(q3),  # variance of yaw angle
+        #     q4 # variance of velocity
+        # ]) ** 2  # predict state covariance
+        # # Observation x,y position covariance
+       # self.R = npediag([r1, r1, r3]) ** 2
+        # self.P = np.eye(4)
+        self.Q = np.diag([
+            0.1,  # variance of location on x-axis
+            0.1,  # variance of location on y-axis
+            np.deg2rad(3),  # variance of yaw angle
+            0.1 # variance of velocity
+        ]) ** 2  # predict state covariance
+        # Observation x,y position covariance
+        self.R = np.diag([0.0, 0.0, 0.3]) ** 2
+        self.P = np.eye(4)
+
+
+
+
+        #BEST SO FAR:
+        # self.Q = np.diag([
+        #     0.2,  # variance of location on x-axis
+        #     0.2,  # variance of location on y-axis
+        #     np.deg2rad(5),  # variance of yaw angle
+        #     0.1 # variance of velocity
+        # ]) ** 2  # predict state covariance
+        # # Observation x,y position covariance
+        # self.R = np.diag([2.0, 2.0, 1.0]) ** 2
+        # self.P = np.eye(4)
+
+
+    def angle_diff(self,ref, act):
```

Comment:
Like anyone coming in here blind will probably be lost at what these methods are actually doing